system/jhtree/jhinplace.cpp

```diff
@@ -2005,7 +2055,17 @@ bool CInplaceLeafWriteNode::add(offset_t pos, const void * _data, size32_t size,
     bool hasSpace = (required <= maxBytes);
     if ((keyLen != keyCompareLen) && ctx.compressionHandler)
     {
-        // Non-Payload size + compression-version byte + optional length (assume the worst)
+        // The following approach is used for compressing payloads:
+        // Payloads are gathered uncompressed until there is not enough space to add the next row.
+        // At that point we try compressing the data.  If it fits when compressed, and the compressed size is
+        // smaller than the uncompressed by a threshold (default 95%) then switch to a compressed payload.
+        // Currently we also stop gathering the uncompressed payload, but that may change in later versions.
+        //
+        // This approach means compression is avoided if the data is small enough to fit uncompressed (e.g. a very
+        // small index), the last leaf node in an index or the payload is not compressible - e.g. mapping to a unique
+        // id.
```

Comment:
Worth mentioning that this means compressors that don't support incremental compression well, now support it slightly better (but not perfectly).